File: cpp_longest_way/src/longest_way.cpp
#include "longest_way.hpp"
#include <iostream>
#include <fstream>
#include <sstream>
#include <stdexcept>


void skip_space(std::string& str) {
	for (std::string::iterator it = str.begin(); it != str.end(); ++it) {
		if (isspace(*it)) {
			str.erase(it);
		}
	}
}

std::vector<edge_t> validate_map_format(const std::string& input_map) {
	std::vector<edge_t> validated_data;
	std::set<std::pair<int, int>> added_edges;
	std::istringstream iss(input_map);
	std::string line;

	while (getline(iss, line)) {
		std::istringstream lineStream(line);
		std::string token;
		std::vector<std::string> tokens;

		while (getline(lineStream, token, ',')) {
			skip_space(token);
			tokens.push_back(token);
		}

		if (tokens.size() != 3) {
			std::cerr << "Invalid map format: too many inputs in one line\n";
			throw std::invalid_argument("Invalid format");
		}

		int start, end;
		double dist;

		if (!std::isdigit(tokens[0][0]) || !std::isdigit(tokens[1][0])) {
			std::cerr << "Invalid map format: ID must be integers\n";
			throw std::invalid_argument("Invalid ID format");
		}

		start = std::stoi(tokens[0]);
		end = std::stoi(tokens[1]);

		if (tokens[0].length() > 7 || tokens[1].length() > 7) {
			std::cerr << "Invalid map format: ID is too large\n";
			throw std::invalid_argument("ID too large");
		}

		// 例外を投げる
		dist = std::stod(tokens[2]);

		if (added_edges.find({start, end}) == added_edges.end()) {
			validated_data.push_back({start, end, dist});
			added_edges.insert({start, end});
		} else {
			std::cerr << "Invalid map format: same IDs exist\n";
			throw std::invalid_argument("Duplicate edge");
		}
	}
	return validated_data;
}

std::vector<edge_t> load_map(const std::string& filename) {
	std::ifstream file(filename);
	if (!file.is_open()) {
		std::cerr << "Error: File '" << filename << "' not found.\n";
		throw std::runtime_error("File not found");
	}

	std::string lines, line;
	while (getline(file, line)) {
		lines += line + "\n";
	}

	return validate_map_format(lines);
}

Graph create_graph(const std::vector<edge_t>& edges) {
	Graph graph;
	for (const auto& edge : edges) {
		add_edge(edge.start, edge.end, boost::property<boost::edge_weight_t, double>(edge.distance), graph);
	}
	return graph;
}

void write_graph_to_dot(const Graph& graph, const std::string& filename) {
	std::ofstream dot_file(filename);
	auto weight_map = get(boost::edge_weight, graph);
	edge_writer<decltype(weight_map)> writer(weight_map);

	boost::write_graphviz(dot_file, graph, boost::default_writer(), writer);
}

// DFSを実行するための補助関数
void dfs(const Graph& graph, Vertex current, std::set<Vertex>& visited, double current_length, std::vector<Vertex>& current_path, double& max_length, std::vector<Vertex>& longest_path) {
	visited.insert(current);
	current_path.push_back(current);

	bool is_leaf = true;
	// 変数名を変更
	auto edges = out_edges(current, graph);
	for (auto ei = edges.first; ei != edges.second; ++ei) {
		Vertex neighbor = target(*ei, graph);
		if (visited.find(neighbor) == visited.end()) {
			is_leaf = false;
			double weight = get(boost::edge_weight, graph, *ei);
			dfs(graph, neighbor, visited, current_length + weight, current_path, max_length, longest_path);
		}
	}

	if (is_leaf && current_length > max_length) {
		max_length = current_length;
		longest_path = current_path;
	}

	visited.erase(current);
	current_path.pop_back();
}

// グラフの全ノードからDFSを開始して最長パスを見つける
std::pair<std::vector<Vertex>, double> find_longest_path_in_graph(const Graph& graph) {
	double max_length = 0;
	std::vector<Vertex> longest_path;

	std::set<Vertex> visited;
	std::vector<Vertex> current_path;
	VertexIter vi, vi_end;
	for (boost::tie(vi, vi_end) = vertices(graph); vi != vi_end; ++vi) {
		visited.clear();
		current_path.clear();
		dfs(graph, *vi, visited, 0, current_path, max_length, longest_path);
	}

	return {longest_path, max_length};
}

File: src/longest_way.py
import networkx as nx
import matplotlib.pyplot as plt
import sys
import signal


def signal_handler(sig, frame):
    print('Ctrl+C pressed')
    sys.exit(127)


def create_graph(validated_data):
    graph = nx.DiGraph()
    for start, end, distance in validated_data:
        graph.add_edge(start, end, weight=distance)
    return graph


# 双方向の場合、距離が描画されるのは片方のみ
def draw_graph(graph):
    pos = nx.spring_layout(graph)
    edge_labels = nx.get_edge_attributes(graph, 'weight')
    nx.draw(graph, pos, with_labels=True, node_color='skyblue', node_size=2000, edge_color='k', linewidths=1,
            font_size=15)
    nx.draw_networkx_edge_labels(graph, pos, edge_labels={e: f"{w:.2f}" for e, w in edge_labels.items()})
    plt.show()


def find_longest_path_in_graph(graph):
    # 深さ優先探索
    def dfs(node, visited, path_length, path):
        visited.add(node)
        max_length = path_length
        max_path = path + [node]
        for neighbor, data in graph[node].items():
            if neighbor not in visited:
                current_length, current_path = dfs(neighbor, visited, path_length + data['weight'], path + [node])
                # 最長片道切符の旅なので距離が同じなものがあった場合、一応経由ポイントが長い方を保持
                if current_length > max_length or (current_length == max_length and len(current_path) > len(max_path)):
                    max_length = current_length
                    max_path = current_path
        visited.remove(node)
        return max_length, max_path

    longest_path = []
    max_distance = float('-inf')
    for node in graph.nodes():
        distance, path = dfs(node, set(), 0, [])
        if distance > max_distance or (distance == max_distance and len(path) > len(longest_path)):
            max_distance = distance
            longest_path = path
    return longest_path, max_distance


# 数値は9999999を最大値に設定
def validate_map_format(input_map):
    validated_data = []
    added_edges = set()
    for line in input_map.splitlines():
        token = line.split(',')
        if len(token) != 3:
            print('Invalid map format: too many inputs in one line')
            return None
        start, end, distance = token[0].strip(), token[1].strip(), token[2].strip()
        if not (start.isdecimal() and end.isdecimal()):
            print('Invalid map format: ID must be integers')
            return None
        if 7 < len(start) or 7 < len(end):
            print('Invalid map format: ID is too large')
            return None
        try:
            dist = float(distance)
            if distance.isdecimal() and 7 < len(distance):
                print('Invalid map format: distance is too large')
                return None
            if not distance.isdecimal() and (7 < len(distance.split('.')[0]) or 2 < len(distance.split('.')[1])):
                print('Invalid map format: part of distance is too large')
                return None
        except ValueError:
            print(f'Invalid map format: {distance} is not a float')
            return None
        edge = (int(start), int(end))
        if edge not in added_edges:
            validated_data.append((int(start), int(end), dist))
            added_edges.add(edge)
        else:
            print('Invalid map format: same IDs exist')
            return None
    return validated_data


def load_map(filename):
    try:
        with open(filename, 'r') as f:
            lines = f.read()
        validated_data = validate_map_format(lines)
        if validated_data is None:
            return None
        return validated_data
    except FileNotFoundError:
        print(f"Error: File '{filename}' not found.")
        return None
    except Exception as e:
        print(f"An error occurred: {str(e)}")
        return None


## 実行速度の目安
# node: 4, edge: 5
# real: 0m0.545s
# user: 0m0.290s
# sys: 0m0.070s
# node: 20, edge: 80
# real: 0m8.497s
# user: 0m8.240s
# sys: 0m0.090s
# node: 80, edge: 160
# real: 0m55.058s
# user: 0m54.410s
# sys: 0m0.330s


def main():
    debug_mode = False
    if '--debug' in sys.argv:
        debug_mode = True
        signal.signal(signal.SIGINT, signal_handler)
        sys.argv.remove('--debug')

    if len(sys.argv) != 2:
        print('Error: one input file needed')
        return
    filename = sys.argv[1]
    validated_data = load_map(filename)
    if validated_data is None:
        return
    graph = create_graph(validated_data)
    longest_path, max_distance = find_longest_path_in_graph(graph)

    for node in longest_path:
        print(node)

    if debug_mode:
        print("Maximum Distance:", max_distance)
        draw_graph(graph)


if __name__ == "__main__":
    main()
